File: 17. Letter Combinations of a Phone Number/17. Letter Combinations of a Phone Number.py
class Solution(object):
    def letterCombinations(self, digits):
        """
        :type digits: str
        :rtype: List[str]
        """

        if  not digits:
            return []
        map = [' ','1','abc','def','ghi','jkl','mno','pqrs','tuv','wxyz']
        res=['']
        for digit in digits:
            resTemp = []
            for s in res:
                for c in map[int(digit)]:
                    resTemp.append(s+c)
            res = resTemp
        return res

File: 17. Letter Combinations of a Phone Number/17. Letter Combinations of a Phone Number.java

import java.util.*;


public class Solution {
    public static List<String> letterCombinations(String digits) {
        LinkedList<String> ans = new LinkedList<String>();
        if(digits == null || digits.length() == 0){
            return ans;
        }
        String[] mapping = new String[] {" ", "1", "abc", "def", "ghi", "jkl", "mno", "pqrs", "tuv", "wxyz"};
        ans.add("");
        for(int i =0; i<digits.length();i++){
            int x = Character.getNumericValue(digits.charAt(i));
            while(ans.peek().length()==i){
                String t = ans.remove();
                for(char s : mapping[x].toCharArray())
                    ans.add(t+s);
            }
        }
        return ans;
    }
    public static void main(String arg[]) {
        System.out.println(letterCombinations("23"));
        System.out.println(letterCombinations("07"));
        System.out.println(letterCombinations("67"));
        System.out.println(letterCombinations("61"));
    }
}
